sdk/core/azure-core/src/http/curl/curl.cpp

```diff
@@ -84,14 +84,48 @@ int pollSocketUntilEventOrTimeout(
   return WSAPoll(&poller, 1, timeout);
 #endif
 }
+
+#ifdef WINDOWS
+// Windows needs this after every write to socket or peformance would be reduced to 1/4 for
+// uploading operation.
+// https://github.com/Azure/azure-sdk-for-cpp/issues/644
+void WinSocketSetBuffSize(curl_socket_t socket, std::function<void(std::string const& message)> logger)
+{
+  ULONG ideal;
+  DWORD ideallen;
+  // WSAloctl would get the ideal size for the socket buffer.
+  if (WSAIoctl(socket, SIO_IDEAL_SEND_BACKLOG_QUERY, 0, 0, &ideal, sizeof(ideal), &ideallen, 0, 0)
+      == 0)
+  {
+    // if WSAloctl succeded (returned 0), set the socket buffer size.
+    // Specifies the total per-socket buffer space reserved for sends.
+    // https://docs.microsoft.com/en-us/windows/win32/api/winsock/nf-winsock-setsockopt
+    auto result = setsockopt(socket, SOL_SOCKET, SO_SNDBUF, (const char*)&ideal, sizeof(ideal));
+    logger("Windows - calling setsockopt after uploading chunk. ideal = " + std::to_string(ideal) + " result = " + std::to_string(result));
```

Comment:
Why do we use `m_logger` here and not `LogThis`?

Reply:
I did m_logger a general function that we can pass to a CurlSession to be called for logging.
Then `LogThis` is created in Details:: namespace to be used for the curlSession that gets created.

This is a way of enabling changing the logging function later or allowing a curlSession to be created without a logging function at all